File: app/modules/plans/schemas.py
from __future__ import annotations

from datetime import datetime

from app.modules.shared.schemas import DashboardModel


class PlanRoleProgress(DashboardModel):
    role: str
    total_checkpoints: int
    done_checkpoints: int


class OpenSpecPlanSummary(DashboardModel):
    slug: str
    title: str
    status: str
    updated_at: datetime
    roles: list[PlanRoleProgress]


class OpenSpecPlansResponse(DashboardModel):
    entries: list[OpenSpecPlanSummary]


class OpenSpecPlanRoleDetail(DashboardModel):
    role: str
    total_checkpoints: int
    done_checkpoints: int
    tasks_markdown: str
    checkpoints_markdown: str | None


class OpenSpecPlanDetail(DashboardModel):
    slug: str
    title: str
    status: str
    updated_at: datetime
    summary_markdown: str
    checkpoints_markdown: str
    roles: list[OpenSpecPlanRoleDetail]

File: apps/frontend/src/features/plans/schemas.ts
import { z } from "zod";

export const PlanRoleProgressSchema = z.object({
  role: z.string().min(1),
  totalCheckpoints: z.number().int().min(0),
  doneCheckpoints: z.number().int().min(0),
});

export const OpenSpecPlanSummarySchema = z.object({
  slug: z.string().min(1),
  title: z.string().min(1),
  status: z.string().min(1),
  updatedAt: z.string().datetime({ offset: true }),
  roles: z.array(PlanRoleProgressSchema).default([]),
});

export const OpenSpecPlansResponseSchema = z.object({
  entries: z.array(OpenSpecPlanSummarySchema).default([]),
});

export const OpenSpecPlanRoleDetailSchema = z.object({
  role: z.string().min(1),
  totalCheckpoints: z.number().int().min(0),
  doneCheckpoints: z.number().int().min(0),
  tasksMarkdown: z.string(),
  checkpointsMarkdown: z.string().nullable(),
});

export const OpenSpecPlanDetailSchema = z.object({
  slug: z.string().min(1),
  title: z.string().min(1),
  status: z.string().min(1),
  updatedAt: z.string().datetime({ offset: true }),
  summaryMarkdown: z.string(),
  checkpointsMarkdown: z.string(),
  roles: z.array(OpenSpecPlanRoleDetailSchema).default([]),
});

export type OpenSpecPlanSummary = z.infer<typeof OpenSpecPlanSummarySchema>;
export type OpenSpecPlansResponse = z.infer<typeof OpenSpecPlansResponseSchema>;
export type OpenSpecPlanDetail = z.infer<typeof OpenSpecPlanDetailSchema>;
